### PR TITLE
S6.3 openai key validation

### LINKED ISSUE #109: S6.3 Add validation for OpenAI key or gracefully handle invalid errors
Three issues:
- [x] We need to implement API key validation.
- [x] We currently only support gpt-4, we need to make this customizable somehow. One quick solution is to check the user's key - if they have 4 let's use 4, otherwise use 3.5. Alternate solution would be to let user pick in their settings
- [x] OpenAI key needs to have full permissions. We should factor that in with the validation + document that.

### PR DESCRIPTION
**Pull request by anthony2261**
*Issue #, if available:*
Closes #109

*Description of changes:*
- Validate API key by calling OpenAI's models endpoint with the key
- Create "preferred_model" user field
- Set the preferred_model field whenever a user create/update call happens with an API key
- Use preferred_model field or default field (new in app config) when calling OpenAI
- Show api key permissions notice on FE

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: text2sql-backend/dataline/services/settings.py [hunk 1]
```diff
@@ -62,10 +69,28 @@ async def update_user_info(self, session: AsyncSession, data: UserUpdateIn) -> U
...
             user = await self.user_repo.create(session, user_create)
         else:
             # Update user with data
             user_update = UserUpdate.model_construct(**data.model_dump(exclude_none=True))
+            if user_update.openai_api_key:
+                key_to_check = user_update.openai_api_key
+                model_to_check = (
+                    user_update.preferred_openai_model or user_info.preferred_openai_model or config.default_model
+                )
+                assert model_exists(
```
Comment (RamiAwar) [inline text2sql-backend/dataline/services/settings.py:87]: We should not use asserts in production code. 

IN the future we may build a binary in OPTIMIZED mode which removes all asserts. 

Just use an if statement instead!

File: text2sql-backend/tests/api/test_settings.py
Comment (RamiAwar) [inline text2sql-backend/tests/api/test_settings.py:1]: Nicee +1 for adding tests

File: text2sql-frontend/src/components/Settings/OpenAIKeyPopup.tsx [hunk 2]
```diff
@@ -56,6 +56,19 @@ export function OpenAIKeyPopup() {
           onKeyUp={handleKeyPress}
         />
       </AlertBody>
+      <AlertDescription>
+        <p className="text-xs">
+          * Your API key must have{" "}
```
Comment (RamiAwar) [inline text2sql-frontend/src/components/Settings/OpenAIKeyPopup.tsx:61]: I would replace this with a call to action:

"Please update your API key with <full permissions> to use DataLine".

As it is now, some users might thing ah ok I dont HAVE to do this.

File: text2sql-frontend/src/components/Settings/Settings.tsx [hunk 3]
```diff
@@ -181,6 +181,17 @@ export default function Account() {
                           onChange={setApiKey}
                         />
                       </div>
+                      <p className="text-xs text-gray-400 pt-2">
```
Comment (RamiAwar) [inline text2sql-frontend/src/components/Settings/Settings.tsx:184]: Same here

File: text2sql-backend/dataline/services/settings.py (hunk 1 above)
Comment (anthony2261) [reply text2sql-backend/dataline/services/settings.py]: True, I thought about the assertion in prod thing but I didn't think of us turning the code into binary in the future. Changes applied 👍

File: text2sql-frontend/src/components/Settings/OpenAIKeyPopup.tsx (hunk 2 above)
Comment (anthony2261) [reply text2sql-frontend/src/components/Settings/OpenAIKeyPopup.tsx]: Done

**Review by RamiAwar — APPROVED** [review]

**Pull request merged**